### PR TITLE
pharo: old -> 10.0.5

### PR DESCRIPTION
**Pull request by jthulhu**
###### Description of changes

The Pharo packages were waaaaay outdated (~10 years old). This is a first attempt to update the packages.

 - [ ] Pharo launcher: the launcher has been deleted from the repo, but hasn't been updated, because the version previously packaged is not compatible with the newly packaged Pharo (at least, I haven't been able to make it work)
 - [ ] Pharo VMs:
   - [X] Spur 64: this has been packaged to version 10.0.5 (which is the latest version available, not to be confused with the version of the Pharo image).
   - [ ] Spur 32: this hasn't been packaged; it *shouldn't* be hard (just changing the source url and the checksum), but it hasn't been done.
   - [X] Cog 32: this is obsolete, so it has been entirely removed.  
 
For this reason, the availability of the Pharo packaged has regressed (only linux x86_64 is supported in this PR).

While this builds and works, any attempt at using Cairo will throw an exception. This should be fixed.

**Important**: the previous maintainer has been removed from the maintainer list, since this packaged was clearly not maintained anymore.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/pharo/default.nix [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  meta = with lib; {
+    description = "Clean and innovative Smalltalk-inspired environment";
+    homepage = "https://pharo.org";
+    longDescription = ''
+      Pharo's goal is to deliver a clean, innovative, free open-source
+      Smalltalk-inspired environment. By providing a stable and small core
+      system, excellent dev tools, and maintained releases, Pharo is an
+      attractive platform to build and deploy mission critical applications.
+
+      This package provides the executable VM. You should probably not care
```
Comment (Mindavi) [inline pkgs/development/pharo/default.nix:89]: Is this part of the description relevant? Also, bug reports should go to the maintainer first before going upstream :)

File: pkgs/development/pharo/default.nix [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    "-DBUILD_BUNDLE=OFF"
+  ];
+
+  installPhase = ''
+    cmake --build . --target=install
+    mkdir -p "$out/lib"
+    mkdir "$out/bin"
+    cp build/vm/*.so* "$out/lib/"
+    cp build/vm/pharo "$out/bin/.pharo"
+    cat >"$out/bin/pharo" <<EOF
```
Comment (Mindavi) [inline pkgs/development/pharo/default.nix:61]: This should be possible using wrapProgram / makeWrapper. See the nixpkgs manual and other derivations for some info

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, left me know if this is clear enough.

File: pkgs/development/pharo/default.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    homepage = "https://pharo.org";
+    longDescription = ''
+      Pharo's goal is to deliver a clean, innovative, free open-source
+      Smalltalk-inspired environment. By providing a stable and small core
+      system, excellent dev tools, and maintained releases, Pharo is an
+      attractive platform to build and deploy mission critical applications.
+    '';
+    license = licenses.mit;
+    maintainers = [ ];
+    platforms = [ "x86_64-linux" ];
```
Comment (drupol) [inline pkgs/development/pharo/default.nix:97]:
```suggestion
    platforms = lib.platforms.linux;
```

**Review by drupol — CHANGES_REQUESTED** [review]
Thanks for implementing the changes.

Last thing, could you sort alphabetically the lambda parameters, the buildInputs, nativeBuildInputs and the meta attributes ?

Don't forget to squash your commits so that this PR only contains 1 single commit.

File: pkgs/development/pharo/default.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+, freetype
+, gcc
+, git
+, gnumake
+, lib
+, libffi
+, libgit2
+, libpng
+, libuuid
+, makeWrapper
```
Comment (drupol) [inline pkgs/development/pharo/default.nix:13]:
```suggestion
, makeBinaryWrapper
```

File: pkgs/development/pharo/default.nix [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  nativeBuildInputs = [
+    cmake
+    freetype
+    gcc
+    git
+    gnumake
+    libffi
+    libuuid
+    makeWrapper
```
Comment (drupol) [inline pkgs/development/pharo/default.nix:57]:
```suggestion
    makeBinaryWrapper
```

**Review by drupol — CHANGES_REQUESTED** [review]
I forgot one last nit.

I usually recommend to use `makeBinaryWrapper` by default since it has a much better compatibility with Darwin platforms.

**Review by drupol — COMMENTED** [review]
LGTM now.

I see that you removed a bunch of entries (pharo-vms, pharo-cog32, etc etc)... shoudln't you add something in `aliases.nix` to let the users know that they doesn't exist anymore? And eventually a message in the release note ?

**Comment by drupol** [comment]
Here's [an example](https://github.com/NixOS/nixpkgs/pull/237824/commits/8a02abacab074e9c2a9ca8af64faeb703d893dac) on how to add an entry in the release note.

The file you need to edit is `nixos/doc/manual/release-notes/rl-2311.section.md`

**Comment by drupol** [comment]
Could you please rebase your PR so there's only one commit?

**Comment by RaitoBezarius** [comment]
> @drupol I'm not sure if I properly rebased, now it looks like it wants to merge 68 commits... I merged from `master`, then I rebased to move my commit to the end.

You failed your rebase and pinged other people, please remove those commits.

**Comment by RaitoBezarius** [comment]
> I'm not too sure how to resolve this merge. If I simply merge, that makes 2 commits and git refuses to amend the previous commit. If I try to rebase, it forcefully inserts commits from master into my branch.

Read the CONTRIBUTING guide on mass pings and retargetting, it is relevant to the problem I imagine.

**Comment by drupol** [comment]
Hi,

Here's a couple of steps that should help you fixing the current issue:

0. `cd /home/.../nixpkgs`
1. `git checkout master`
2. `git remote add upstream git@github.com:NixOS/nixpkgs.git`
3. `git fetch --all`
4. `git reset upstream/master --hard`
5. `git push`
6. `git checkout update-pharo`
7. `git rebase master`
8. `git push -f`

**Review by drupol — APPROVED** [review]
LGTM !

**Pull request merged**